### PR TITLE
Migrates android-api and adds integration api.

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Also migrates geolocation. 

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
hey @latin-panda . Mind having a look? Thank you!

File: webapp/src/ts/services/geolocation.service.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  private stopWatching() {
+    console.debug('Cancelling geolocation');
+    this.watcher && window.navigator.geolocation && window.navigator.geolocation.clearWatch(this.watcher);
+  }
+
+  private defer() {
+    this.deferred = {};
+    this.deferred.promise  = new Promise((resolve, reject) => {
```
Comment (latin-panda) [inline webapp/src/ts/services/geolocation.service.ts:93]:
```suggestion
    this.deferred.promise = new Promise((resolve, reject) => {
```

File: webapp/tests/karma/ts/services/android-api.service.spec.ts [hunk 2]
```diff
@@ -1,35 +1,36 @@
...
+import sinon from 'sinon';
+import { expect } from 'chai';
+
+import { AndroidApiService } from '@mm-services/android-api.service';
+import { SessionService } from '@mm-services/session.service';
+import { GeolocationService } from '@mm-services/geolocation.service';
+import { FeedbackService } from '@mm-services/feedback.service';
 
-  'use strict';
+describe('AndroidApi service', () => {
```
Comment (latin-panda) [inline webapp/tests/karma/ts/services/android-api.service.spec.ts:10]: Is it possible or worth adding unit test for any of these:
-  back
- logout
- mrdtResponse
- mrdtTimeTakenResponse
- simprintsResponse
- smsStatusUpdate
- locationPermissionRequestResolved

File: webapp/tests/karma/ts/services/geolocation.service.spec.ts [hunk 3]
```diff
@@ -137,13 +124,14 @@ describe('Geolocation service', () => {
...
         speed: 7,
       };
-      $window.navigator.geolocation.watchPosition.callsFake((success, failure) => {
+      // @ts-ignore
+      window.navigator.geolocation.watchPosition.callsFake((success, failure) => {
         success({coords: first});
         failure({code: 2, message: 'some later error'});
       });
 
       return service.init()().then(returned => {
```
Comment (latin-panda) [inline webapp/tests/karma/ts/services/geolocation.service.spec.ts:133]: For this one and the following unit test:
I think it will look better if these functions are in separate line.
Also I tend to like better async / await to for unit test but these still work fine, something optional

**Review by latin-panda — COMMENTED** [review]
Nice!! Looks good, just minor comments.
Btw I saw we're using route snapshot, then I guess we don't need the must updated route state or it doesn't apply ?

**Comment by dianabarsan** [comment]
> Btw I saw we're using route snapshot,

I'm not using routeSnapshot, I just used that variable name.
You can't subscribe to snapshot in services, well, you can but you don't get events. 
I'm actually using:
```
let route = this.router.routerState.root;
    while (route.firstChild) {
      route = route.firstChild;
    }
    return route.snapshot;
````
inspired from: https://github.com/angular/angular/issues/15004#issuecomment-412525030

File: webapp/tests/karma/ts/services/geolocation.service.spec.ts (hunk 3 above)
Comment (dianabarsan) [reply webapp/tests/karma/ts/services/geolocation.service.spec.ts:133]: I just copied and pasted these tests and just fixed what was broken. Would you mind if I just leave them as they were?

File: webapp/src/ts/app.module.ts [hunk 4]
```diff
@@ -94,4 +94,11 @@ export class MissingTranslationHandlerLog implements MissingTranslationHandler {
   ],
   bootstrap: [AppComponent]
 })
-export class AppModule { }
+export class AppModule {
+  integration;
+  constructor(
+    private integrationApiService:IntegrationApiService,
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/app.module.ts:100]: So this was really weird. I couldn't access constructor parameters directly, even if they were not private, though I didn't set them as explicitly "public".

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/app.module.ts:100]: Oh that's fine :)

File: webapp/tests/karma/ts/services/geolocation.service.spec.ts (hunk 3 above)
Comment (latin-panda) [reply webapp/tests/karma/ts/services/geolocation.service.spec.ts:133]: That’s okay, it's the original :)

File: webapp/src/ts/services/android-api.service.ts [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    }
+    try {
+      response = JSON.parse(response);
+    } catch(e) {
+      return console.error(new Error('Unable to parse JSON response from android app: "' + response + '"'));
+    }
+    if (requestType === 'identify') {
+      //Simprints.identifyResponse(requestId, response);
+    } else if (requestType === 'register') {
+      //Simprints.registerResponse(requestId, response);
```
Comment (craig-landry) [inline webapp/src/ts/services/android-api.service.ts:239]: @dianabarsan  I see the TODO up on line 25 for the simprints service, which I assume is why these `Simprints` lines are commented out. Did you forget to come back to this, or will this be handled in a separate PR?

File: webapp/src/ts/services/android-api.service.ts [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    }
+  }
+
+  smsStatusUpdate(id, destination, content, status, detail) {
+    console.debug('smsStatusUpdate() :: ' +
+      ' id=' + id +
+      ', destination=' + destination +
+      ', content=' + content +
+      ', status=' + status +
+      ', detail=' + detail);
```
Comment (craig-landry) [inline webapp/src/ts/services/android-api.service.ts:251]:
```suggestion
    console.debug(`smsStatusUpdate() ::  id=${id}, destination=${destination}, content=${content}, status=${status}, detail=${detail}`);
```

I know this block was already formatted like this, but an alt-enter and "replace with template string" in webstorm makes it a bit more pleasant :)

File: webapp/src/ts/services/integration-api.service.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import { TranslateService } from '@ngx-translate/core';
+
+import { LanguageService } from '@mm-services/language.service';
+import { Select2SearchService } from '@mm-services/select2-search.service';
+import { MRDTService } from '@mm-services/mrdt.service';
+import { MarkdownService } from '@mm-services/markdown.service';
+import { SettingsService } from '@mm-services/settings.service';
+import { AndroidApiService } from '@mm-services/android-api.service';
+import { DbService } from '@mm-services/db.service';
+// todo migrate enketo
```
Comment (craig-landry) [inline webapp/src/ts/services/integration-api.service.ts:11]: Similar to my question above about the `simprints` TODO, is this to be handled in a separate PR?

**Review by craig-landry — COMMENTED** [review]
Comments are non-blocking. Mostly seeking to understand.

File: webapp/src/ts/services/android-api.service.ts (hunk 5 above)
Comment (dianabarsan) [reply webapp/src/ts/services/android-api.service.ts:239]: Simprints service is currently being migrated by @njogz

File: webapp/src/ts/services/integration-api.service.ts (hunk 7 above)
Comment (dianabarsan) [reply webapp/src/ts/services/integration-api.service.ts:11]: As above.

Comment (dianabarsan) [reply webapp/src/ts/services/integration-api.service.ts:11]: Actually, my bad 👍🏻 it's not as above. 

I'm intentionally splitting the Contact summary + Integration api + Enketo into different PRs to make it more manageable. 
So, it will be handled into a future PR.

File: webapp/src/ts/services/markdown.service.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      const $childStore = $('<div/>');
+      $(this).children(':not(input, select, textarea)').each(function(index) {
+        const name = '$$$' + index;
+        $service.translateElement($(this).clone()).appendTo($childStore);
+        $(this).replaceWith(name);
+      });
+
+      html = $service.basic($(this).html());
+
+      $childStore.children().each(function(i) {
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/services/markdown.service.ts:54]: These need to be `function`s, so they have their own context and resolve `this` correctly.

File: webapp/src/ts/services/geolocation.service.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+@Injectable({
+  providedIn: 'root'
+})
+export class GeolocationService {
+  private readonly GEO_OPTIONS = {
+    enableHighAccuracy: true,
+    timeout: 30 * 1000, // give up if coords not received within 30 seconds
+    maximumAge: 5 * 60 * 1000 // coords from up to 5 minutes ago are acceptable
+  }
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/services/geolocation.service.ts:11]: I really need to setup tslint!!!

File: webapp/src/ts/services/android-api.service.ts (hunk 6 above)
Comment (dianabarsan) [reply webapp/src/ts/services/android-api.service.ts:251]: Unfortunately, if I do this, the line exceeds the allowed length. We don't have tslint enabled yet, so nothing complains, but it'll need reformatting when we enable it. 
For reference, these are our eslint rules: https://github.com/medic/eslint-config/blob/master/index.js#L12

File: webapp/tests/karma/ts/services/android-api.service.spec.ts (hunk 2 above)
Comment (dianabarsan) [reply webapp/tests/karma/ts/services/android-api.service.spec.ts:13]: I added a bunch of tests, for each of these. The `back` is super tricky to test because of all the jquery, so I just added tests to cover router navigation.

**Comment by dianabarsan** [comment]
@latin-panda would you mind having another look? Thanks!

**Review by latin-panda — APPROVED** [review]
Excellent! I don't have additional feedback :)
